hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/handlers/EntityHandler.java

```diff
@@ -0,0 +1,366 @@
+    this.omMetadataManager = omMetadataManager;
+    this.reconSCM = reconSCM;
+    this.bucketHandler = bucketHandler;
+  }
+
+  public abstract NamespaceSummaryResponse getSummaryResponse()
+          throws IOException;
+
+  public abstract DUResponse getDuResponse(
+          boolean listFile, boolean withReplica)
+          throws IOException;
+
+  public abstract QuotaUsageResponse getQuotaResponse()
+          throws IOException;
+
+  public abstract FileSizeDistributionResponse getDistResponse()
+          throws IOException;
+
+  public ReconOMMetadataManager getOmMetadataManager() {
+    return omMetadataManager;
+  }
+
+  public OzoneStorageContainerManager getReconSCM() {
+    return reconSCM;
+  }
+
+  public ReconNamespaceSummaryManager getReconNamespaceSummaryManager() {
+    return reconNamespaceSummaryManager;
+  }
+
+  public BucketHandler getBucketHandler() {
+    return bucketHandler;
+  }
+
+  public String getNormalizedPath() {
+    return normalizedPath;
+  }
+
+  public String[] getNames() {
+    return names.clone();
+  }
+
+  /**
+   * Return the entity handler of client's request, check path existence.
+   * If path doesn't exist, return UnknownEntityHandler
+   * @param reconNamespaceSummaryManager ReconNamespaceSummaryManager
+   * @param omMetadataManager ReconOMMetadataManager
+   * @param reconSCM OzoneStorageContainerManager
+   * @param path the original path request used to identify root level
+   * @return the entity handler of client's request
+   */
+  public static EntityHandler getEntityHandler(
+          ReconNamespaceSummaryManager reconNamespaceSummaryManager,
+          ReconOMMetadataManager omMetadataManager,
+          OzoneStorageContainerManager reconSCM,
+          String path) throws IOException {
+    BucketHandler bucketHandler;
+    normalizedPath = normalizePath(path);
+    names = parseRequestPath(normalizedPath);
+
+    if (path.equals(OM_KEY_PREFIX)) {
+      return EntityType.ROOT.create(reconNamespaceSummaryManager,
+              omMetadataManager, reconSCM, null);
+    }
+
+    if (names.length == 0) {
+      return EntityType.UNKNOWN.create(reconNamespaceSummaryManager,
+              omMetadataManager, reconSCM, null);
+    } else if (names.length == 1) { // volume level check
+      String volName = names[0];
+      if (!volumeExists(omMetadataManager, volName)) {
+        return EntityType.UNKNOWN.create(reconNamespaceSummaryManager,
+                omMetadataManager, reconSCM, null);
+      }
+      return EntityType.VOLUME.create(reconNamespaceSummaryManager,
+              omMetadataManager, reconSCM, null);
+    } else if (names.length == 2) { // bucket level check
+      String volName = names[0];
+      String bucketName = names[1];
+      OmBucketInfo omBucketInfo = OmBucketInfo.newBuilder()
+          .setVolumeName(volName)
+          .setBucketName(bucketName)
+          .setBucketLayout(BucketLayout.FILE_SYSTEM_OPTIMIZED)
```

Comment:
Why is this being set to FSO?  The call that was here before seemed correct:
```
      omBucketInfo = OzoneManagerUtils
              .getOmBucketInfo(omMetadataManager, volName, bucketName);
```

Reply:
It needs to be initialized and by `getOmBucketInfo()` we are getting a null value. The only way to initialize it is by using `newBuilder()` which has a default value of LEGACY for bucket layout. Setting it to FSO is just a workaround for now. There were some test methods failing because of this but now they are passing. I'll have to figure out a better solution later.